Alert App Skeleton (CU-d8qufg)

trisapeace opened the pull request:
Whew, that took awhile. I think I'm finally ready to get some feedback on the Alert App skeleton.

**Still to do:**
* Add webhooks for server-to-app communication.
* Try it out on a couple of real devices (I'll probably need help with the iPhone testing because I think all the test devices that I have access to are iPhone 5 or 5c which means they are too old).
* New formatting/linting based on the outcome of https://app.clickup.com/t/eprhhn .

**What I'm looking for:**
* Feedback on pretty much everything, but particularly in style, organization, and choice to tools/libraries.
* Where it would be useful to have more documentation to explain what's going on or how something should be done.
* Whether it can be run locally by @schwarrrtz and @rhoberman . 

**How to go review this PR:**
* I have a lot of information in the commit messages about what I'm doing and why. And I've tried to keep the relevant parts in each commit, so it might be easiest to go through the PR that way instead of being overwhelmed by it all at once. But it's totally up to you.
* I'm going to book a meeting where we can continue to discuss this and/or troubleshoot any problems you're having getting it running locally.

rhoberman commented on src/components/ErrorScreen/ErrorScreen.jsx:
```diff
@@ -0,0 +1,47 @@
+    View,
+} from 'react-native'
+import { get } from 'lodash'
+import colors from '../../resources/colors'
```
If there is any chance that we will want to support theming? Apple does require that we support Dark and Light modes (although we can use a single colour scheme if it works for both modes.)

If we can't rule this out then I would consider the approach to theming used in BeSafe. It doesn't add much in the way of complexity but it facilitates (and hopefully encourages) a maintainble approach. That said, I'm busy refactoring it so it should look a lot better soon.

https://developer.apple.com/documentation/xcode/supporting_dark_mode_in_your_interface?language=objc

trisapeace replied:
Cool. I didn't realize that it was going to be required for iOS. Then yes, I think I should put it in. When do you think you'll be done your refactoring? It would probably be best to wait for that, right?

This `colors.js` structure was also from the BeSafe codebase, so is it not being used there?

trisapeace replied:
@rhoberman Any update on the Theme refactoring?

rhoberman commented on src/components/HistoricAlert/HistoricAlert.jsx:
```diff
@@ -0,0 +1,103 @@
+        flex: 1,
+        flexDirection: 'row',
+    },
+    timeText: {
+        fontSize: 20,
+    },
```
I would encourage defining typography in a single place. This makes it easier to:

- ensure consistency
- tweak typographic choices
- detect proliferation of typography
- verify that the implementation and design system are aligned

Did you find the approach in BeSafe overly complex?

trisapeace replied:
I started copying it over, but it felt like overkill when I don't actually know which set of fonts are going to be used in the design or how many I'd need (you have Paragraph, and then Heading1 through Heading8. That felt like a lot). I figured that it could be put into place once the first real screens start being implemented.

I also got confused that `Paragraph` used `Heading2`. And that's when I stopped.

trisapeace replied:
Leaving this until the Theme refactor: https://app.clickup.com/t/k93bw9

rhoberman commented on src/components/HistoricAlert/HistoricAlert.stories.jsx:
```diff
@@ -0,0 +1,31 @@
+    .add('Urgent Alert', () => (
+        <HistoricAlert
+            roomName='1st Floor Landing'
+            category='Other'
```
I think we already discussed this and decided it could easily be deferred, but in case we didn't, BeSafe is enjoying the react-i18next library and it is relatively simple to integrate.

rhoberman replied:
Just realised I'm commenting on text in a story so please read this in the context of actual components.

trisapeace replied:
Yes, I think we can leave the internationalization out for now. Although that reminds me to ask Kenzie to try for English that is good in both Canada and the US when she writes the copy 😄

rhoberman commented on src/screens/ExampleScreen2.jsx:
```diff
@@ -0,0 +1,118 @@
+            />
+            <Button
+                title="Start alarm"
+                onPress={() => startAlarm()}
```
I belive an exception here will blow-up your app (from memory React will unmount, but it has been a while). Recommend that all of your event handlers catch and handle exceptions. This was a key rationale for the safeHandler - it makes it easy to create robust error handlers.

trisapeace replied:
Gooood point. Thanks! I'll add that.

rhoberman commented on src/services/SoundService.js:
```diff
@@ -0,0 +1,43 @@
+    }
+})
+
+function startAlarm() {
+    alarm.play((success) => {
+        if (success) {
+            logger.debug('Successfully finished playing alarm')
+        } else {
+            logger.debug('Playback of alarm failed due to audio decoding errors')
+        }
+    })
+    logger.debug('Started playing alarm')
+}
```
What happens if the invocation results in an error? Does it get handled/reported? (I'm running out of time so I haven't looked at the Sound API.

trisapeace replied:
Hmm, I'm not exactly sure. I would hope that it would call the callback with a false-y value. However, just in case, I've wrapped the invocation of this function in a try-catch.

trisapeace commented:
Task linked: [CU-d8qufg Skeleton version of the app front end works locally](https://app.clickup.com/t/d8qufg)

trisapeace commented:
@rhoberman @schwarrrtz @risatronic @mariocimet 
Is anyone else interested in looking at this again before I merge it? I'm feeling anxious to move on from this and start building the real app on top of this skeleton.

### New since last time:
* Very basic Web Sockets. Lots will need to be added to this to make it robust when we actually start using it.
* Prettier/ESLint
* Moved test files into `/test` 
* Some changes to get the iOS build to deploy on BitRise (e.g. added fake app icon image files)
* Uses Jest instead of Mocha for tests (which fixes the awkward mocking problem with `react-native-device-info` that we talked about before)
* BraveAlertService is now split up into ButtonsApiService and SensorApiService, so that FetchService can be a generic HTTP fetcher without any of our business logic

The server side of these test POST and WebSocket endpoints can be found here: https://github.com/bravetechnologycoop/BraveButtons/tree/td-app-backend
I have that code currently up on `chatbot-dev` so that the functions will work when I run them locally. But neither of these functions will continue to exist after a real call of each type is implemented.

schwarrrtz reviewed (approved):
I can build and run this locally on the iOS simulator after some shenanigans involving `npm install -g react-native-cli` and `pod install`. The React Native docs don't seem to approve of the global CLI installation, but I wasn't able to get it to work otherwise.

I think the best thing now would be to merge this and start building on it, and I'll maybe wipe my install and try again at some point now that I know more about how it works. I'll submit a separate pull request with changes to the docs covering what I had to do to get it working on my machine.
